### PR TITLE
feat(CI): Add python virtualenvs to github caches

### PR DESCRIPTION
**Pull request by szalai1**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: .github/workflows/metadata-ingestion.yml [hunk 1]
```diff
@@ -55,8 +55,13 @@ jobs:
...
         run: ./metadata-ingestion/scripts/install_deps.sh
-      - name: Run metadata-ingestion tests (extras ${{ matrix.extraPythonRequirement }})
-        run: ./gradlew -Pextra_pip_requirements='${{ matrix.extraPythonRequirement }}' :metadata-ingestion:${{ matrix.command }}
+      - uses: actions/cache@v3
+        id: cache-venv
+        with:
+          path: ./metadata-ingestion/venv/
+          key: ${{ runner.os }}-venv-${{ hashFiles('./metadata-ingestion/setup.py') }}-${{ matrix.python-version }}-${{ matrix.command }}
+      - name: Run metadata-ingestion tests
+        run: ./gradlew :metadata-ingestion:build :metadata-ingestion:${{ matrix.command }} -x:metadata-ingestion:testQuick -x:metadata-ingestion:check
```
Comment (hsheth2) [inline .github/workflows/metadata-ingestion.yml:64]: why are we tweaking the gradle command here?

File: .github/workflows/metadata-ingestion.yml [hunk 2]
```diff
@@ -55,8 +55,13 @@ jobs:
           hadoop-version: "3.2"
       - name: Install dependencies
         run: ./metadata-ingestion/scripts/install_deps.sh
-      - name: Run metadata-ingestion tests (extras ${{ matrix.extraPythonRequirement }})
-        run: ./gradlew -Pextra_pip_requirements='${{ matrix.extraPythonRequirement }}' :metadata-ingestion:${{ matrix.command }}
+      - uses: actions/cache@v3
+        id: cache-venv
+        with:
+          path: ./metadata-ingestion/venv/
+          key: ${{ runner.os }}-venv-${{ hashFiles('./metadata-ingestion/setup.py') }}-${{ matrix.python-version }}-${{ matrix.command }}
```
Comment (hsheth2) [inline .github/workflows/metadata-ingestion.yml:62]: does this need restore keys?

File: .github/workflows/metadata-ingestion.yml [hunk 3]
```diff
@@ -55,8 +55,13 @@ jobs:
...
+      - name: Run metadata-ingestion tests
```
Comment (szalai1) [inline .github/workflows/metadata-ingestion.yml:63]:
```suggestion
      - name: Run metadata-ingestion tests (extras ${{ matrix.extraPythonRequirement }})
```

File: .github/workflows/metadata-ingestion.yml (hunk 1 above)
Comment (szalai1) [inline .github/workflows/metadata-ingestion.yml:64]:
```suggestion
        run: ./gradlew -Pextra_pip_requirements='${{ matrix.extraPythonRequirement }}' :metadata-ingestion:${{ matrix.command }}
```

File: .github/workflows/docker-unified.yml [hunk 4]
```diff
@@ -414,6 +414,13 @@ jobs:
         if: ${{ needs.setup.outputs.publish != 'true' }}
         with:
           image: ${{ env.DATAHUB_MAE_CONSUMER_IMAGE }}:${{ needs.setup.outputs.unique_tag }}
+      - uses: actions/cache@v3
+        id: cache-venv
+        with:
+          path: ./smoke-test/venv/
+          key: smoke-venv-${{ runner.os }}-${{ hashFiles('./smoke/requirements.txt') }}
+          restore-keys: |
```
Comment (szalai1) [inline .github/workflows/docker-unified.yml:422]:
```suggestion
```

File: .github/workflows/docker-unified.yml [hunk 5]
```diff
@@ -414,6 +414,13 @@ jobs:
...
+            smoke-venv-${{ runner.os }}-
```
Comment (szalai1) [inline .github/workflows/docker-unified.yml:423]:
```suggestion
```

File: .github/workflows/metadata-ingestion.yml (hunk 2 above)
Comment (szalai1) [reply .github/workflows/metadata-ingestion.yml:62]: More like the other one doesn't need it. Removed it there.

Comment (szalai1) [reply .github/workflows/metadata-ingestion.yml:62]: If it can find it by key, then that's a good match any other changes causes to rebuild the cache.

Comment (hsheth2) [reply .github/workflows/metadata-ingestion.yml:62]: makes sense

**Review by hsheth2 — APPROVED** [review]

**Comment by hsheth2** [comment]
@szalai1 I was looking into this a bit more and I don't think this is the right way to implement caching. I think we might be able to use the example under "Caching projects that use setup.py" from https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#caching-packages, but not 100% that it'll work with our matrix setup since it doesn't let us set the cache key ourselves.

More broadly, we want to cache pip's cache directory, but not the `venv` itself. The issue with caching the venv directory is that it can cause pip to leave outdated packages as-is instead of downloading and installing the newer version of the package. That means that a build without caching could differ from a build with the cache enabled.

**Comment by szalai1** [comment]
@hsheth2 thanks for review. so these are valid concerns. I think the main "problem" with `setup-python` is that it caches the downloads and not the installs, which is the majority of the time. 

A solution to not leaving outdated installs in venv would be: 
1. `pip install -r requirments.txt --dry-run > would_be_installed.txt`
2. use `hashFiles('./would_be_installed.txt') }` for caching 

This way we can be sure that we invalidate the cached `venv` whenever a new package is available.

**Comment by hsheth2** [comment]
@szalai1 yup that should work. One thing to be careful of is that we still always want to run the `pip install -e .` step.

**Comment by szalai1** [comment]
@hsheth2 Added dry run hashes to the venv cache name to trigger rebuild if anything would change. 

`install -e .` will run everytime, even if the venv is resotored. 

This saved 7mins for me locally for smoke-tests.

**Pull request closed without merging**